lib/util/backoff.js

```diff
@@ -0,0 +1,102 @@
+
+/**
+ * Expose `Backoff`.
+ */
+
+class Backoff {
+  /**
+   * Construct a {@link Backoff}.
+   * @param {function} fn - Function to call
+   * @param {object} options
+   * @property {number} min - Initial timeout in milliseconds [100]
+   * @property {number} max - Max timeout [10000]
+   * @property {boolean} jitter - Apply jitter [0]
+   * @property {number} factor - Multiplication factor for Backoff operation [2]
+   */
+  constructor(fn, options) {
+    Object.defineProperties(this, {
+      _min: {
+        value: options.min || 100
+      },
+      _max: {
+        value: options.max || 10000
+      },
+      _jitter: {
+        value: options.jitter > 0 && options.jitter <= 1 ? options.jitter : 0
+      },
+      _factor: {
+        value: options.factor || 2
+      },
+      _attempts: {
+        value: 0,
+        writable: true
+      },
+      _duration: {
+        enumerable: true,
+        get() {
+          let ms = this._min * Math.pow(this._factor, this._attempts);
+          if (this._jitter) {
+            const rand =  Math.random();
+            const deviation = Math.floor(rand * this._jitter * ms);
+            ms = (Math.floor(rand * 10) & 1) === 0  ? ms - deviation : ms + deviation;
+          }
+          return Math.min(ms, this._max) | 0;
+        }
+      },
+      _fn: {
+        value: fn || null,
+        writable: true
+      },
+      _timeoutID: {
+        value: null,
+        writable: true
+      },
+    });
+  }
+
+  /**
+  * Increases attempts and calls the function.
+  *
+  * @return {void}
+  * @api public
+  */
+  onTimer() {
+    this._attempts++;
+    this._fn();
+    if (this._attempts !== 0) {
+      this.backoff();
+    }
+  }
+
+  /**
+  * Start the backoff operation.
+  *
+  * @return {void}
+  * @api public
+  */
+  backoff(fn) {
```

Comment:
i didn't understand why we need `fn` parameter, once we have it in constructor.  We can talk more about it offline